### ISSUE TITLE
Switch subset with Loader

### ISSUE THREAD

**Issue by aardschok** [getavalon/core#310]
## Goal
Allow user to change the subset of a loaded asset.

## Motivation
A situation has arrived which can perfectly describe the support for this feature.
We want to automate the step from render to slap comp. This will require the following things:

* Rendered image sequences
* A predefined comp (created per project)
* A logic to update the subset / version / representation of any Loader in the comp

The current demands for this logic are that it will need to be able to clear a container from data and re-imprint it with new data. This will result in the new subset being loaded with the latest version, key here is __latest__.
The logic will only be able to use the Loader from the existing container to update the subset/version/representation, this will also keep us away from headaches and over complex functions.

It will look something along the lines of this, function in the `api.Loader` of `core/avalon/pipeline` for Fusion:

```python
# PSEUDO FUNCTION OF api.Loader
def switch(container, representation):
    from avalon.fusion import comp_lock_and_undo_chunk, imprint_container, clean_container
    
    host =  api.registered_host()
    tool = container["_tool"]
    assert tool.ID == "Loader", "Must be Loader"
    comp = tool.Comp()

    root = api.get_representation_path(representation)
    context = api.get_representation_context(representation)
    path = self._get_first_image(root)
    
    
    namespace = context['asset']['name']
    subset = context["subset"]["name"]
    version = context["version"]
    family = version["data"]["families"][0]

    # Ensure this subset can be loaded
    assert family in self.families, "Unsupported family for loader `%s`" % self.label

    clean_container(tool)  # here we strip all the attributes of it's values
    imprint_container(tool,
                                subset,
                                namespace,
                                context,
                                self.__class__.__name__
```

## Other uses
This will also allow the user to swap different subsets of let's say, cameras. Instead of creating a new camera version for stereo cameras the user can simply swap a `cameraDefault` subset for `cameraStereo` or `stereoCamDefault`, assuming they are using the same Loader.

Another example is swapping Level of Detail models of the same character. Let's say the camera is so close to an object that you will want to update it to the high resolution. `modelLOD1` is loaded but we need `modelLOD0` (lod0 being closest to the cam and lodX the most distant) 

## Conclusion
Implement swapping of subset when it meets these requirements:

* Family supported by Loader used
* Representation supported by Loader used

This will mean a new function to strip containers of their values.

**Comment by BigRoy** [comment]
> The logic will only be able to use the Loader from the existing container to update the subset/version/representation, this will also keep us away from headaches and over complex functions.

To me this sentence is somewhat confusing. But I imagine this refers to the fact that one *container* loaded with `ModelLoader` can't switch to a subset with a different loader. It can only switch to another subset's version that is supported by the Loader's families.

---

To me the implementation details sound good. 

I do believe though that the `switch` method itself doesn't need to "ensure" it's a valid switch, similar to how the `load` also doesn't validate it's actually triggering on a correct representation. The call to it already did so in `avalon.pipeline`, so there's `avalon.pipeline.load` and `avalon.pipeline.update`, and now there's `avalon.pipeline.switch`.

Which would be:

```python
def switch(container, representation_id):
    """Switch a container to representation"""

    # Compute the different version from 'representation'
    current_representation = io.find_one({
        "_id": io.ObjectId(container["representation"])
    })
    assert current_representation is not None, "This is a bug"

    # Get the Loader for this container
    Loader = _get_container_loader(container)
    if not Loader:
        raise RuntimeError("Can't switch container. See log for details.")

    if not hasattr(Loader, "switch"):
        raise RuntimeError("Loader does not support 'switch'")

    # Get the new representation to switch to
    new_representation = io.find_one({
        "type": "representation",
        "_id": representation_id,
    })

    new_context = get_representation_context(new_representation)
    assert is_compatible_loader(Loader, new_context), "Must be compatible Loader"

    Loader = _make_backwards_compatible_loader(Loader)
    loader = Loader(get_representation_context(container["representation"]))
    return loader.switch(container, new_representation)
```

Likely most Loaders can already "switch" by just calling `self.update(container, new_representation)`

**Comment by mottosso** [comment]
I like the sound of this. :)

**Comment by BigRoy** [comment]
Implemented with #312

![](https://camo.githubusercontent.com/47a0cfd94a0f142d6b771ddd70de62f7c111f2e5/68747470733a2f2f692e6779617a6f2e636f6d2f31313266313262346136633234613662653833373234643362306633393539662e676966)
